python-threatexchange/threatexchange/signal_type/signal_base.py

```diff
@@ -13,25 +13,44 @@
 from threatexchange.signal_type import index
 
 
-class HashComparisonResult(t.NamedTuple):
+class SignalComparisonResult(t.NamedTuple):
     match: bool
-    distance: int  # TODO - remove
+    distance: index.SignalSimilarityInfo
 
     @classmethod
-    def from_match(cls, dist: int = 0) -> "HashComparisonResult":
-        return cls(True, dist)
+    def from_bool_only(cls, matches: bool) -> "SignalComparisonResult":
+        """For SignalTypes with no distance"""
+        return cls.from_match_only() if matches else cls.from_no_match_only()
 
     @classmethod
-    def from_no_match(cls, dist: int = 1) -> "HashComparisonResult":
-        return cls(False, dist)
+    def from_match_only(cls) -> "SignalComparisonResult":
+        """For SignalTypes with no distance"""
+        return cls(True, index.SignalSimilarityInfo())
 
     @classmethod
-    def from_dist(cls, dist: int, threshold: int) -> "HashComparisonResult":
-        return cls(dist <= threshold, dist)
+    def from_no_match_only(cls) -> "SignalComparisonResult":
+        """For SignalTypes with no distance"""
+        return cls(False, index.SignalSimilarityInfo())
 
     @classmethod
-    def from_bool(cls, matches: bool) -> "HashComparisonResult":
-        return cls.from_match() if matches else cls.from_no_match()
+    def from_simple_dist(
+        cls, dist: index.CT, threshold: index.CT
+    ) -> "SignalComparisonResult":
+        """For SignalTypes with simple distance"""
+        return cls(
+            dist <= threshold,
+            index.SignalSimilarityInfoWithSingleDistance[index.CT](dist),
+        )
+
+    @classmethod
+    def from_dist(
+        cls, dist: index.SignalSimilarityInfo, threshold: index.SignalSimilarityInfo
+    ) -> "SignalComparisonResult":
+        """For SignalTypes w"""
+        return cls(
+            dist <= threshold,
```

Comment:
unsure: SignalSimilarityInfo has override options for  `__lt__` and `__eq__` 
Is python structured in such a way (i.e. smart enough) that `__le__` will respect these changes to `__lt__` + `__eq__` ?

Reply:
<= will only call `__le__`, the implementation for `lt` + `eq` won't be used. I could `t.final` `__le__` to force it to be based on the above. 

You can use `functools.total_ordering` to define all the operators based on a strictly sortable type with only two operators (lt + eq) defined, which is why those are the ones I forced to be implemented